File: report/src/main/java/com/hartwig/actin/report/pdf/tables/EligibleActinTrialsGenerator.java
package com.hartwig.actin.report.pdf.tables;

import java.util.List;
import java.util.Set;
import java.util.StringJoiner;

import com.google.common.collect.Lists;
import com.hartwig.actin.algo.datamodel.TreatmentMatch;
import com.hartwig.actin.algo.interpretation.EvaluatedTrial;
import com.hartwig.actin.algo.interpretation.EvaluatedTrialExtractor;
import com.hartwig.actin.report.pdf.util.Cells;
import com.hartwig.actin.report.pdf.util.Formats;
import com.hartwig.actin.report.pdf.util.Tables;
import com.itextpdf.layout.element.Table;

import org.apache.logging.log4j.util.Strings;
import org.jetbrains.annotations.NotNull;

public class EligibleActinTrialsGenerator implements TableGenerator {

    @NotNull
    private final List<EvaluatedTrial> trials;
    @NotNull
    private final String source;
    private final float trialIdColWidth;
    private final float acronymColWidth;
    private final float cohortColWidth;
    private final float molecularEventColWidth;
    private final float criteriaToCheckColWidth;
    private final float mainCheckColWidth;
    private final float isOpenColWidth;


    @NotNull
    public static EligibleActinTrialsGenerator fromTreatmentMatch(@NotNull TreatmentMatch treatmentMatch, @NotNull String source,
            float contentWidth) {
        float trialIdColWidth = contentWidth / 10;
        float acronymColWidth = contentWidth / 10;
        float cohortColWidth = contentWidth / 10;
        float molecularColWidth = contentWidth / 10;
        float criteriaToCheckColWidth = contentWidth / 10;
        float isOpenColWidth = contentWidth / 10;
        float mainCheckColWidth =
                contentWidth - (trialIdColWidth + acronymColWidth + cohortColWidth + molecularColWidth + criteriaToCheckColWidth
                        + isOpenColWidth);

        List<EvaluatedTrial> trials = EvaluatedTrialExtractor.extract(treatmentMatch);
        List<EvaluatedTrial> filtered = Lists.newArrayList();
        for (EvaluatedTrial trial : trials) {
            if (trial.isPotentiallyEligible()) {
                filtered.add(trial);
            }
        }
        return new EligibleActinTrialsGenerator(filtered,
                source,
                trialIdColWidth,
                acronymColWidth,
                cohortColWidth,
                molecularColWidth,
                criteriaToCheckColWidth,
                mainCheckColWidth,
                isOpenColWidth);
    }

    public EligibleActinTrialsGenerator(@NotNull final List<EvaluatedTrial> trials, @NotNull final String source,
            final float trialIdColWidth, final float acronymColWidth, final float cohortColWidth, final float molecularEventColWidth,
            final float criteriaToCheckColWidth, final float mainCheckColWidth, final float isOpenColWidth) {
        this.trials = trials;
        this.source = source;
        this.trialIdColWidth = trialIdColWidth;
        this.acronymColWidth = acronymColWidth;
        this.cohortColWidth = cohortColWidth;
        this.molecularEventColWidth = molecularEventColWidth;
        this.criteriaToCheckColWidth = criteriaToCheckColWidth;
        this.mainCheckColWidth = mainCheckColWidth;
        this.isOpenColWidth = isOpenColWidth;
    }

    @NotNull
    @Override
    public String title() {
        return source + " trials and cohorts considered potentially eligible (" + eligibleCount(trials) + ")";
    }

    private static int eligibleCount(@NotNull List<EvaluatedTrial> trials) {
        int eligibleCount = 0;
        for (EvaluatedTrial trial : trials) {
            if (trial.isPotentiallyEligible()) {
                eligibleCount++;
            }
        }
        return eligibleCount;
    }

    @NotNull
    @Override
    public Table contents() {
        Table table = Tables.createFixedWidthCols(trialIdColWidth,
                acronymColWidth,
                cohortColWidth,
                molecularEventColWidth,
                criteriaToCheckColWidth,
                mainCheckColWidth,
                isOpenColWidth);

        table.addHeaderCell(Cells.createHeader("Trial ID"));
        table.addHeaderCell(Cells.createHeader("Acronym"));
        table.addHeaderCell(Cells.createHeader("Cohort"));
        table.addHeaderCell(Cells.createHeader("Molecular event"));
        table.addHeaderCell(Cells.createHeader("# Criteria to check"));
        table.addHeaderCell(Cells.createHeader("# Main check"));
        table.addHeaderCell(Cells.createHeader("Cohort open"));

        for (EvaluatedTrial trial : trials) {
            table.addCell(Cells.createContent(trial.trialId()));
            table.addCell(Cells.createContent(trial.acronym()));
            table.addCell(Cells.createContent(trial.cohort() != null ? trial.cohort() : Strings.EMPTY));
            table.addCell(Cells.createContent("TODO"));
            table.addCell(Cells.createContent(String.valueOf(trial.evaluationsToCheckCount())));
            table.addCell(Cells.createContent(concat(trial.evaluationsToCheckMessages())));
            table.addCell(Cells.createContentYesNo(trial.isOpen() ? "Yes" : "No"));
        }

        return Tables.makeWrapping(table);
    }

    @NotNull
    private static String concat(@NotNull Set<String> strings) {
        StringJoiner joiner = Formats.commaJoiner();
        for (String string : strings) {
            joiner.add(string);
        }
        return Formats.valueOrDefault(joiner.toString(), "None");
    }
}

File: report/src/main/kotlin/com/hartwig/actin/report/pdf/tables/trial/EligibleActinTrialsGenerator.kt
package com.hartwig.actin.report.pdf.tables.trial

import com.hartwig.actin.report.interpretation.InterpretedCohort
import com.hartwig.actin.report.pdf.tables.TableGenerator
import com.hartwig.actin.report.pdf.tables.trial.ActinTrialGeneratorFunctions.addTrialsToTable
import com.hartwig.actin.report.pdf.tables.trial.ActinTrialGeneratorFunctions.createTableTitleStart
import com.hartwig.actin.report.pdf.util.Cells
import com.hartwig.actin.report.pdf.util.Tables
import com.hartwig.actin.report.pdf.util.Tables.makeWrapping
import com.hartwig.actin.report.trial.filterCohortsAvailable
import com.itextpdf.layout.element.Table

class EligibleActinTrialsGenerator(
    private val cohorts: List<InterpretedCohort>,
    private val title: String,
    private val footNote: String?,
    private val trialColWidth: Float,
    private val cohortColWidth: Float,
    private val molecularEventColWidth: Float,
    private val locationColWidth: Float?,
    private val checksColWidth: Float
) : TableGenerator {

    private val includeLocation = locationColWidth != null

    override fun title(): String {
        return title
    }

    override fun contents(): Table {

        val table = Tables.createFixedWidthCols(
            trialColWidth, cohortColWidth + molecularEventColWidth + (locationColWidth ?: 0f) + checksColWidth
        )
        val widths = listOfNotNull(cohortColWidth, molecularEventColWidth, locationColWidth, checksColWidth).toFloatArray()

        if (cohorts.isNotEmpty()) {
            table.addHeaderCell(Cells.createContentNoBorder(Cells.createHeader("Trial")))
            val headerSubTable = Tables.createFixedWidthCols(*widths)
            sequenceOf("Cohort", "Molecular", "Hospitals".takeIf { includeLocation }, "Warnings").filterNotNull()
                .map(Cells::createHeader).forEach(headerSubTable::addHeaderCell)
            table.addHeaderCell(Cells.createContentNoBorder(headerSubTable))
        }
        addTrialsToTable(
            cohorts = cohorts,
            table = table,
            tableWidths = widths,
            feedbackFunction = InterpretedCohort::warnings,
            includeLocation = includeLocation
        )
        if (footNote != null) {
            table.addCell(Cells.createSpanningSubNote(footNote, table))
        }
        return makeWrapping(table)
    }

    fun getCohortSize(): Int {
        return cohorts.size
    }

    companion object {

        fun forOpenCohorts(
            cohorts: List<InterpretedCohort>, source: String?, width: Float, slotsAvailable: Boolean, includeLocation: Boolean = false
        ): EligibleActinTrialsGenerator {
            val recruitingAndEligibleCohorts = filterCohortsAvailable(cohorts, slotsAvailable)
            val recruitingAndEligibleTrials = recruitingAndEligibleCohorts.map(InterpretedCohort::trialId).distinct()
            val slotsText = if (!slotsAvailable) " but currently have no slots available" else ""
            val cohortFromTrialsText = if (recruitingAndEligibleCohorts.isNotEmpty()) {
                "(${formatCountWithLabel(recruitingAndEligibleCohorts.size, "cohort")} " +
                        "from ${formatCountWithLabel(recruitingAndEligibleTrials.size, "trial")})"
            } else {
                "(0)"
            }

            val title = "${createTableTitleStart(source)} that are open and potentially eligible$slotsText $cohortFromTrialsText"

            return create(recruitingAndEligibleCohorts, title, width, null, includeLocation)
        }

        fun forOpenCohortsWithMissingMolecularResultsForEvaluation(
            cohorts: List<InterpretedCohort>, source: String?, width: Float, includeLocation: Boolean = false
        ): EligibleActinTrialsGenerator? {
            val recruitingAndEligibleCohorts = cohorts.filter {
                it.isPotentiallyEligible && it.isOpen && it.isMissingMolecularResultForEvaluation!!
            }
            val recruitingAndEligibleTrials = recruitingAndEligibleCohorts.map(InterpretedCohort::trialId).distinct()
            val cohortFromTrialsText = when {
                recruitingAndEligibleCohorts.isNotEmpty() -> "(${
                    formatCountWithLabel(
                        recruitingAndEligibleCohorts.size,
                        "cohort"
                    )
                }" + " from ${formatCountWithLabel(recruitingAndEligibleTrials.size, "trial")})"

                else -> "(0)"
            }

            val title =
                "${createTableTitleStart(source)} that are open but additional molecular tests needed to evaluate eligibility $cohortFromTrialsText"
            val footNote = "Open cohorts with no slots available are shown in grey."

            return if (recruitingAndEligibleCohorts.isNotEmpty()) create(
                recruitingAndEligibleCohorts, title, width, footNote, includeLocation
            ) else null
        }

        private fun formatCountWithLabel(count: Int, word: String): String {
            return "$count $word${if (count > 1) "s" else ""}"
        }

        fun forClosedCohorts(
            cohorts: List<InterpretedCohort>, source: String?, contentWidth: Float, includeLocation: Boolean = false
        ): EligibleActinTrialsGenerator {
            val unavailableAndEligible = cohorts.filter { trial: InterpretedCohort -> trial.isPotentiallyEligible && !trial.isOpen }
            val title =
                "${createTableTitleStart(source)} and cohorts that are potentially eligible, but are closed (${unavailableAndEligible.size})"
            return create(unavailableAndEligible, title, contentWidth, null, includeLocation)
        }

        private fun create(
            cohorts: List<InterpretedCohort>, title: String, width: Float, footNote: String? = null, includeLocation: Boolean
        ): EligibleActinTrialsGenerator {
            val trialColWidth = width / 9
            val cohortColWidth = width / 4
            val molecularColWidth = width / 7
            val locationColWidth = if (includeLocation) width / 7 else 0f
            val checksColWidth = width - (trialColWidth + cohortColWidth + molecularColWidth + locationColWidth)
            return EligibleActinTrialsGenerator(
                cohorts,
                title,
                footNote,
                trialColWidth,
                cohortColWidth,
                molecularColWidth,
                locationColWidth.takeIf { includeLocation },
                checksColWidth
            )
        }
    }
}